### PR TITLE
fix a bug on determining srid when upload a vector layer

### LINKED ISSUE #289: Layer map view does not zoom to layer correctly
A newly uploaded layer shows in the layer list fine but it does not zoom correctly to the layer:

http://geonode.kartoza.com/layers/geonode:tandale_imagery

screen shot 2017-07-01 at 10 26 53 am

child of #457

### LINKED ISSUE #435: Layer Upload not working for Vector Layers
When we upload vector layer, the EPSG code didn't get read. So, srid of the layer were saved as 'EPSG:None'.

The problem might be very deep in EPSG detection on GeoNode. We might have to file bugfix and unittest on upstream.

child of #457

### PR DESCRIPTION
**Pull request by boney-bun**
fix #289 
fix #435

### CODE REVIEW

**Comment by boney-bun** [comment]
This PR is addressing what i believe as a bug at uploading a vector layer.
The problem lies at determining the layer's srid. 
It returned `EPSG:None`.
In the case of *Africa Volcano* layer, the SRS is:
```
GEOGCS["GCS_WGS_1984",
    DATUM["WGS_1984",
        SPHEROID["WGS_84",6378137,298.257223563]],
    PRIMEM["Greenwich",0],
    UNIT["Degree",0.017453292519943295]]
```
However, the geonode doesn't recognize this.
The vector layer doesn't use *gdal* and *osr* functions to determine the SRS as in raster layer.
So, we expect the SRS to be 4326. 
But, the code failed to do so.
So, this PR is basically make sure that if the layer's SRS can't be determined, then it should be set to 4326.

@gubuntu I understand that we should go with `3857` as the default. But, `4326` is what the upstream do.

the PR in action:
geonode 437

The first `Africa Volcano` illustrates the original behaviour when uploading a vector layer (no thumbnail and can't see the layer detail).
The second one denotes another bug on upstream (can't see the layer detail).
The last illustration indicates the PR in action (no errors).

File: geonode/layers/utils.py [hunk 1]
```diff
@@ -375,7 +377,8 @@ def get_bbox(filename):
         bbox_y0 = min(ext[0][1], ext[2][1])
         bbox_x1 = max(ext[0][0], ext[2][0])
         bbox_y1 = max(ext[0][1], ext[2][1])
-        srid = srs.GetAuthorityCode(None) if srs else 'EPSG:4326'
+        # eliminate default EPSG srid as it will be added when this function returned
+        srid = srs.GetAuthorityCode(None) if srs else '4326'
 
     return [bbox_x0, bbox_x1, bbox_y0, bbox_y1, "EPSG:%s" % str(srid)]
```
Comment (myarjunar) [inline geonode/layers/utils.py]: even an expert geonode developer can miss this :joy:

**Comment by NyakudyaA** [comment]
@boney-bun cc @myarjunar  I think it is wrong to say if ` layer's SRS can't be determined, then it should be set to 4326.`
This would result in problems later because the layer will render in a different place on the globe. I think what you should rather be doing is let the upload fail with an error message when the SRS cannot be determined. 

**NB:** Not all projections are entered in the EPSG database but the important thing is the [projection string](https://proj4.org/parameters.html)

[cape.zip](https://github.com/kartoza/geonode/files/2126646/cape.zip)
I have attached a sample dataset which is in South African local Transverse Mercator - This projection is not added in the EPSG database so you won't find the EPSG code for that but the proj string is still valid.

**Comment by myarjunar** [comment]
@NyakudyaA I reckon what @boney-bun did here is only to fix programming error that causes the method returned `EPSG:None` with the original logic is still following what upstream already did.

**Comment by boney-bun** [comment]
thanks @myarjunar you've got my point.

@NyakudyaA @gubuntu this is where i need an opinion from a GIS expert :)

technically speaking, the original code: 
`srid = layer.srs.srid if layer.srs else '4326'`
is basically checking if `layer.srs` is empty.
in the case of *Africa Volcano* layer, `layer.srs` is not empty.
but when executing `layer.srs.srid`, it returned None.
geonode doesn't recognize the projection info from the vector layer.
in fact, `layer.srs` returned the following information:
(cmiiw, I expect the layer.srs.srid return something like: EPSG:4326 because the below info is equal to WGS84)
```
GEOGCS["GCS_WGS_1984",
    DATUM["WGS_1984",
        SPHEROID["WGS_84",6378137,298.257223563]],
    PRIMEM["Greenwich",0],
    UNIT["Degree",0.017453292519943295]]
```
so to accomodate @NyakudyaA  comment, our options are:
- teach geonode to recognize the above info from the vector layer, or
- use *gdal* and *osr* as in raster layer.
- leave it as it's from the upstream

we may need to make a PR to the upstream.

which one do you prefer @NyakudyaA @gubuntu ?

**Comment by NyakudyaA** [comment]
@boney-bun I think the best way you could do is do a PR against upstream with your changes. 
I think you might need to consider the following:
* How does the EPSG: Code gets used in the front end to render the layer.
* Instead of using the EPSG: Code can you use `layer.srs.proj4`. This should cover all cases where the EPSG code is not defined in the projection file.
* I know when using gdal to project layers to different formats you can use either the EPSG code or the proj4 string and still the results will generate the layer in the same position.

**Comment by boney-bun** [comment]
Hi @NyakudyaA 

I've changed the code a bit.
It seems that gdal's `DataSource` is unreliable to determine the epsg code (the function reads the shapefile).
So, I change to reading a prj file and it can now detect the epsg code properly.
This PR seems to partially solve #395 too as illustrated below:
uploadlayer-1

For `4326` it can also be uploaded successfully:
uploadlayer-2

@NyakudyaA @gubuntu please review the code from the GIS perspective. then, i could make a pr to the upstream.

File: geonode/layers/utils.py (hunk 1 above)
Comment (lucernae) [reply geonode/layers/utils.py]: This should have been tested by unittest. If not, then we should add.

Comment (lucernae) [reply geonode/layers/utils.py]: Also we need to submit PR to upstream GeoNode for issues like this.

**Comment by boney-bun** [comment]
thanks for the review @NyakudyaA @myarjunar @lucernae 
it has passed the travis so I'm going to merge them.

File: geonode/layers/utils.py (hunk 1 above)
Comment (gubuntu) [reply geonode/layers/utils.py]: @boney-bun cc @lucernae we should be doing all GeoNode PRs directly upstream.

Also why is this merged when we decided to not assume 4326 if no projection defined?

**Pull request merged**